### ISSUE TITLE
Cache invalidation not working with Nx Daemon on gitpod

### ISSUE THREAD

**Issue by dsschneidermann** [nrwl/nx#18866]
### Current Behavior

Nx local cache invalidation seems to be not working. If I change the content of a file listed on the inputs of the task and run the task again, I get an erroneous cache hit and the output of the task doesn't change.

Note that I have copied this repro from a closed issue (#18289), and fixing the error in output path in that reproduction does not prevent the cache miss, as demonstrated by the linked reproduction repo here.

Also here is a screenshot of it failing to rebuild after a change.
![image](https://github.com/nrwl/nx/assets/11848465/eee7355e-d96d-42a3-be8d-f25ca85bf73f)

 `"useDaemonProcess": false` fixes this for me, so I will be keeping that setting permanently from now.

### Expected Behavior

A cache miss would be expected.

### GitHub Repo

https://github.com/dsschneidermann/nx-cache-invalidation-bug-repro

### Steps to Reproduce

1. Clone the repository
2. Run npm install
3. Run npm run test
4. Change the contents of the string inside the console.log in packages/test/src/index.js
5. Run npm run test
6. Compare the outputted string with the expected

### Nx Report

```shell
>  NX   Report complete - copy this into the issue template

   Node   : 18.17.1
   OS     : linux-x64
   npm    : 9.6.7
   
   nx         : 16.7.4
   @nrwl/tao  : 16.7.4
   nx-cloud   : 16.1.1
```

### Operating System

- [ ] macOS
- [X] Linux
- [ ] Windows
- [ ] Other (Please specify)

**Comment by fmqa** [comment]
Still an issue on nx 16.8.1 / node 18.17.1 / npm 10.0.0 - also on Linux. This is reproducible even with small monorepos.
Thankfully, addiing`"useDaemonProcess": false` to the task runner options or setting the environment variable `NX_DAEMON=false` mitigate the issue for the time being.

**Comment by andersonba** [comment]
Wow!! I finally found the issue, thanks @dsschneidermann.
The same thing happens to me. I use runtime inputs to invalidate some caches, and after a few updates, it stopped working, making our dev environment quite unstable.
By running with `env NX_DAEMON=false`, the cache gets invalidated again after repeated attempts. Awesome! 🙏 

Is the fix on the team's radar?

**Comment by SebastienGllmt** [comment]
Ran into the same issue today when setting up `nx` for the first time

My changes to the Nx config would never work driving me insane as I poured over the Nx documentation trying to figure out what I was doing wrong. However, I every time I rage quit and came back to it, it would suddenly work again

After a lot of debugging, I now finally realize the issue is caused by the Daemon who fails to reload any changes to my files. The reason rage-quitting causes it to work again is because the Daemons [restarts after 3hrs of inactivity](https://github.com/nrwl/nx/blob/971a020419734285929f8746cc0c446df66ab761/packages/nx/src/daemon/server/shutdown-utils.ts#L9) so when you come back to the task things magically work

You can fix this by adding `"useDaemonProcess": false` to your `nx.json`

```json
"tasksRunnerOptions": {
    "default": {
      "runner": "nx/tasks-runners/default",
      "options": {
        "useDaemonProcess": false,
      }
    }
  },
```

**Comment by dmitry-stepanenko** [comment]
Here's a repo with slightly different issue that seems related https://github.com/dmitry-stepanenko/nx-named-inputs-invalid-cache . Check the README for reproduction steps

I'm declaring a namedInput `{ "runtime": "node log.js" }` and its being ignored by the Nx Daemon.

**Comment by Cammisuli** [comment]
@dsschneidermann I tried running your repo in a Ubuntu install, and I couldn't reproduce your issue: 
![image](https://github.com/nrwl/nx/assets/4332460/597a250d-c87b-4b3a-a068-ad58aed667c3)

Do things start working properly if you do `npx nx reset`?

**Comment by Cammisuli** [comment]
@dmitry-stepanenko we know about your issue with regards to the runtime input and envs. As a workaround for you, can you try using the [env input](https://nx.dev/reference/project-configuration#env-variables) directly instead of running it through the runtime input?

**Comment by Cammisuli** [comment]
the fix for the runtime inputs using the env is here: [19422](https://github.com/nrwl/nx/pull/19422)

**Comment by dsschneidermann** [comment]
> @dsschneidermann I tried running your repo in a Ubuntu install, and I couldn't reproduce your issue: ![image](https://user-images.githubusercontent.com/4332460/272278597-597a250d-c87b-4b3a-a068-ad58aed667c3.png)
> 
> Do things start working properly if you do `npx nx reset`?

Thanks for looking into this. I suspect it has something to do with the file monitoring of the nx daemon and the way that Gitpod is running their virtual OS. You can replicate it with (a free account on) Gitpod: https://gitpod.io/#https://github.com/dsschneidermann/nx-cache-invalidation-bug-repro

I am not really aware of the underlying complexities here, but it seems clear that the watching fails to get notified. I am aware of plenty of other tools are able to correctly monitor files in this specific virtual environment (eg. `nodemon` comes to mind), so the problem cannot be entirely placed on Gitpods shoulders. I suspect that the issue can also be reproduced locally in a docker-based environment - maybe relating somehow to host mounts and so on. It's certainly annoying that dev environments can vary on such a basic premise, and the fact that the Nx daemon is a default "on" can make it very difficult for some users to get started.

To answer your question, `npx nx reset` helps only once and then behaves the same again, so it doesn't provide any workaround.

**Comment by John98Zakaria** [comment]
I am running the latest version of nx 17.0.1 (node v18.18.2) on ubuntu 22.04 and have the same problem

**Comment by juniperoserra** [comment]
We are seeing the same behavior on using NFS on macOS. Local filesystems work fine, but when we use networked shares on Mac, the watcher in the daemon never triggers. Other applications on Mac that are listening to file change work OK. And the same share on Linux works OK.

Stripped of actual paths, this is what `nfsstat -vm` says for one our our mount points:

```
/System/Volumes/path/to/mount/point from nfsserver:/exportpath
  -- Original mount options:
     General mount flags: 0x10500010 nodev,automounted,noatime,nobrowse
     NFS parameters: proto=tcp,noconn,locallocks,nordirplus,timeo=600,retrans=2
     File system locations:
       /exportpath @ nfsserver (dotquad)
  -- Current mount parameters:
     General mount flags: 0x14500010 nodev,automounted,noatime,nobrowse,multilabel
     NFS parameters: vers=3,proto=tcp,port=2049,nomntudp,hard,nointr,noresvport,negnamecache,callumnt,locallocks,quota,rsize=32768,wsize=32768,readahead=16,dsize=32768,nordirplus,nodumbtimer,timeo=600,maxgroups=16,acregmin=5,acregmax=60,acdirmin=5,acdirmax=60,nomutejukebox,nonfc,sec=sys,readlink_nocache=0
     File system locations:
       /exportpath @ nfsserver (dotquad)
     fh 20 01000600f77929ad00349f0c0000000000000000
     Current location: 0x1 0 0 0: 
       /exportpath @ nfsserver (dotquad)
     Status flags: 0x0
```

**Comment by flippidippi** [comment]
> Ran into the same issue today when setting up `nx` for the first time
> 
> My changes to the Nx config would never work driving me insane as I poured over the Nx documentation trying to figure out what I was doing wrong. However, I every time I rage quit and came back to it, it would suddenly work again
> 
> After a lot of debugging, I now finally realize the issue is caused by the Daemon who fails to reload any changes to my files. The reason rage-quitting causes it to work again is because the Daemons [restarts after 3hrs of inactivity](https://github.com/nrwl/nx/blob/971a020419734285929f8746cc0c446df66ab761/packages/nx/src/daemon/server/shutdown-utils.ts#L9) so when you come back to the task things magically work
> 
> You can fix this by adding `"useDaemonProcess": false` to your `nx.json`
> 
> ```json
> "tasksRunnerOptions": {
>     "default": {
>       "runner": "nx/tasks-runners/default",
>       "options": {
>         "useDaemonProcess": false,
>       }
>     }
>   },
> ```

This fixed it for us also. We are using WSL. It won't recalculate, which causes really confusing issues where nx thinks that a projects hash hasn't changed but it has.

**Comment by vktrl** [comment]
Same issue, disabled the daemon for now.

WSL / Ubuntu 22.04 / Nx 17.2.6

**Comment by danieladugyan** [comment]
+1
WSL2 / Ubuntu 22.04.3 / Nx 17.2.8

**Comment by inorganik** [comment]
I'm experiencing the same issue; the build cache gets used regardless of code changes. Turning cache off under "targetDefaults" > "build" works too:

```js
"targetDefaults": {
    "build": {
      "cache": false,
      "dependsOn": ["^build"],
      "inputs": ["production", "^production"],
    },
```
But isn't build caching a core feature of NX? How could maintainers leave a bug like this for so long?

**Comment by Cammisuli** [comment]
hey all, I'm using this to track the original issue of the daemon not working properly on gitpod. This is known, and Nx doesn't work well in that environment. 

For everything else, can you make a new issue with clear steps, OSes used, environments that are interesting (ie, we test thoroughly on local file systems, what is different about your setup?) so that I can try my best to reproduce it?
